### PR TITLE
Add required TLS for placement service

### LINKED ISSUE #1279: Enable TLS client-server cert validation for placement service if mTLS is enabled
The placement service currently runs without TLS authentication.

While there is no change of sensitive information ever traversing through the placement service, and thus removing the need for in-transit encryption, a malicious code in the cluster or network can become a host for Actors by registering itself with the placement service, and thus starting to receive actor activation requests.

When running with mTLS, requiring the gRPC server for the placement service to authenticate with client-server credentials would only allow authenticated Dapr sidecars to register themselves as legitimate hosts.

### PR DESCRIPTION
**Pull request by yaron2**
Closes #1279 

This PR does the following:

1. Adds mandatory TLS for placement service if mTLS is enabled
2. Refactors common logic into `credentials` package for reuse
3. Adds and updates tests
4. Moves `tls.enabled` setting to `global` in the Helm Chart

### CODE REVIEW

File: pkg/actors/actors.go [hunk 1]
```diff
@@ -535,18 +541,45 @@ func (a *actorsRuntime) connectToPlacementService(placementAddress, hostAddress
 	}()
 }
 
+func (a *actorsRuntime) getClientOptions(certChain *dapr_credentials.CertChain) ([]grpc.DialOption, error) {
```
Comment (amanbha) [inline pkg/actors/actors.go]: can this also be moved to credentials/grpc.go

File: cmd/placement/main.go [hunk 2]
```diff
@@ -28,6 +33,8 @@ func main() {
 	metricsExporter := metrics.NewExporter(metrics.DefaultMetricNamespace)
 	metricsExporter.Options().AttachCmdFlags(flag.StringVar, flag.BoolVar)
 
+	flag.StringVar(&certChainPath, "certchain", "/var/run/dapr/credentials", "Path to the credentials directory holding the cert chain")
```
Comment (amanbha) [inline cmd/placement/main.go]: Can this be moved as const to crednetials pkg, its used at multiple places in codebase in operator, sentry etc.

File: pkg/actors/actors.go (hunk 1 above)
Comment (yaron2) [reply pkg/actors/actors.go]: done!

File: cmd/placement/main.go (hunk 2 above)
Comment (yaron2) [reply cmd/placement/main.go]: for this, I'd rather keep it here as flags are directly tied to the process, and I'd want to keep the option of changing them if needed..

Comment (amanbha) [reply cmd/placement/main.go]: sure, please make it a const string here.

Comment (yaron2) [reply cmd/placement/main.go]: done

File: pkg/credentials/credentials.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
+package credentials
+
+import (
+	"path/filepath"
+)
+
+type TLSCredentials struct {
```
Comment (shalabhms) [inline pkg/credentials/credentials.go]: might need to add comment on public struct, linter wont be happy...

File: pkg/credentials/credentials.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+	credentialsPath string
+}
+
+func NewTLSCredentials(path string) TLSCredentials {
```
Comment (shalabhms) [inline pkg/credentials/credentials.go]: nit : comments ..applicable to all public methods/structs..

File: cmd/operator/main.go [hunk 5]
```diff
@@ -10,6 +10,7 @@ import (
 	"time"
 
 	scheme "github.com/dapr/dapr/pkg/client/clientset/versioned"
+	certchain "github.com/dapr/dapr/pkg/credentials"
```
[the remark this answers is not in the document]
Comment (yaron2) [reply cmd/operator/main.go]: Fixed these import paths

File: cmd/placement/main.go [hunk 6]
```diff
@@ -50,8 +57,38 @@ func main() {
...
+
+		log.Info("mTLS enabled, getting tls certificates")
+		// try to load certs from disk, if not yet there, start a watch on the local filesystem
+		chain, err := credentials.LoadFromDisk(tlsCreds.RootCertPath(), tlsCreds.CertPath(), tlsCreds.KeyPath())
+		if err != nil {
+			fsevent := make(chan struct{})
+
+			go func() {
+				log.Infof("starting watch for certs on filesystem: %s", certChainPath)
+				err = fswatcher.Watch(context.Background(), tlsCreds.Path(), fsevent)
```
[the remark this answers is not in the document]
Comment (yaron2) [reply cmd/placement/main.go]: I don't agree here, as `Path` in the context of TLS Credentials denote the path where the entirety of the credentials are hosted. In addition, there's an explicit comment there for that.

File: cmd/placement/main.go [hunk 7]
```diff
@@ -50,8 +57,38 @@ func main() {
 	stop := make(chan os.Signal, 1)
 	signal.Notify(stop, os.Interrupt)
 
+	var certChain *credentials.CertChain
+	if tlsEnabled {
+		tlsCreds := credentials.NewTLSCredentials(certChainPath)
+
+		log.Info("mTLS enabled, getting tls certificates")
+		// try to load certs from disk, if not yet there, start a watch on the local filesystem
+		chain, err := credentials.LoadFromDisk(tlsCreds.RootCertPath(), tlsCreds.CertPath(), tlsCreds.KeyPath())
```
[the remark this answers is not in the document]
Comment (yaron2) [reply cmd/placement/main.go]: Path is not Certs Path, its a path for the key as well as two other certificates.

File: pkg/credentials/credentials.go (hunk 3 above)
Comment (yaron2) [reply pkg/credentials/credentials.go]: Done

File: charts/dapr/charts/dapr_config/templates/dapr_default_config.yaml [hunk 8]
```diff
@@ -5,7 +5,7 @@ metadata:
   name: default
 spec:
   mtls:
-    enabled: {{ .Values.mtls.enabled }}
+    enabled: {{ .Values.global.mtls.enabled }}
     workloadCertTTL: {{ .Values.mtls.workloadCertTTL }}
```
[the remark this answers is not in the document]
Comment (yaron2) [reply charts/dapr/charts/dapr_config/templates/dapr_default_config.yaml]: Moved to global config, makes sense.

File: charts/dapr/values.yaml [hunk 9]
```diff
@@ -6,4 +6,6 @@ global:
   imagePullSecrets: ""
   prometheus:
     enabled: true
-    port: 9090
\ No newline at end of file
+    port: 9090
+  mtls:
+    enabled: true
```
[the remark this answers is not in the document]
Comment (yaron2) [reply charts/dapr/values.yaml]: Moved, see above comment.

File: cmd/placement/main.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (shalabhms) [reply cmd/placement/main.go]: We can continue to have Path() but then CertPath be changed to IssuerCertPath() ?

File: cmd/placement/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (shalabhms) [reply cmd/placement/main.go]: ..then can we make CertPath to IssuerCertPath() ?

File: pkg/operator/api/api.go [hunk 10]
```diff
@@ -49,13 +46,16 @@ func NewAPIServer(client scheme.Interface) Server {
...
+func (a *apiServer) Run(certChain *dapr_credentials.CertChain) {
 	lis, err := net.Listen("tcp", fmt.Sprintf(":%v", serverPort))
 	if err != nil {
 		log.Fatal("error starting tcp listener: %s", err)
 	}
 
-	opts := a.getServerOptions(certChain)
+	opts, err := dapr_credentials.GetServerOptions(certChain)
+	if err != nil {
+		log.Fatal("error creating gRPC options: %s", err)
```
[the remark this answers is not in the document]
Comment (yaron2) [reply pkg/operator/api/api.go]: there's a creation of an x509 in there, so I wouldn't denote it as getting.

File: cmd/placement/main.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (yaron2) [reply cmd/placement/main.go]: In the future we may change the default path returned to be configurable, and I don't want to tie the generic Cert to an Issuer Cert. The fact that currently it serves an issuer cert is an implementation detail.

File: pkg/credentials/credentials.go (hunk 4 above)
Comment (yaron2) [reply pkg/credentials/credentials.go]: This has comments.

File: pkg/credentials/credentials.go [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+func (t *TLSCredentials) Path() string {
+	return t.credentialsPath
+}
+
+// RootCertPath returns the file path for the root cert
+func (t *TLSCredentials) RootCertPath() string {
+	return filepath.Join(t.credentialsPath, RootCertFilename)
+}
+
+// CertPath returns the file path for the cert
```
[the remark this answers is not in the document]
Comment (yaron2) [reply pkg/credentials/credentials.go]: See my comment above

File: cmd/placement/main.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (shalabhms) [reply cmd/placement/main.go]: Sounds good.

**Pull request merged**